src/Stripe.net/Infrastructure/Public/HttpClient.cs

```diff
@@ -0,0 +1,32 @@
+namespace Stripe
+{
+    using System;
+    using System.Threading;
+    using System.Threading.Tasks;
+
+    /// <summary>
+    /// Abstract base class for HTTP clients used to make requests to Stripe's API.
+    /// </summary>
+    public abstract class HttpClient
+    {
+        /// <summary>Gets or sets the last request made by this client.</summary>
+        /// <value>The last request made by this client.</value>
+        public StripeRequest LastRequest { get; protected set; }
```

Comment:
How does the concept of `LastRequest` work with async requests where multiple requests are ongoing?

I'm asking because I know we want to add telemetry to .NET too but this one is fully tailored for async and sync requests so not sure how that will fit here.

Reply:
`LastRequest` contains the request that was sent last, and `LastResponse` contains the response that was received last. There is no guarantee that the two match, i.e. it's very possible that `LastResponse` contains the response to a different request than the one in `LastRequest`.

For telemetry, we will do something similar to stripe-java and use a [`ConcurrentQueue`](https://docs.microsoft.com/en-us/dotnet/api/system.collections.concurrent.concurrentqueue-1?view=netframework-4.7.2) to enqueue metrics each time a request completes, and dequeue metrics each time a request is sent.

Reply:
@ob-stripe Don't you think that by exposing `LastResponse` / `LastRequest` inline with an asynchronous method that we're exposing a bit of a misleading API? i.e., You can access these properties, but you won't be able to use them reliably for most things.

I might be missing something here, but couldn't we still get the telemetry we need by not exposing them and doing some management of requests and responses down in `StripeClient. MakeRequestAsync`?

Reply:
TBH I cargo-culted `LastRequest` and `LastResponse` from Twilio's SDK -- we don't need them at all for telemetry. I was thinking they could be useful in tests or more generally in contexts where you know what requests are being sent (typically because only one request is sent).

If you think the API could be confusing, I'm fine with making those properties `internal` or even removing them entirely.

Reply:
Ah, I see. My main objection was just that the majority of things people might try to use the properties for wouldn't actually be safe, even if they might have initially seemed safe. I see your point though that it could be useful to have access to this kind of thing during testing (although I think using any form of parallel testing could be potential for trouble).

I'd probably lean toward removal or marking `internal` to avoid the possibility of misuse, but I don't feel super strongly about it. If we do keep them public though, I think we should comprehensively document them to indicate that they're not safe to use in most cases.

Reply:
Ack. I've removed the properties and turned `HttpClient` from an abstract class to an interface (and renamed it to `IHttpClient`).